### ISSUE TITLE
make PythonAPI error : Could not find compiler set in environment variable CC

### ISSUE THREAD

**Issue by GZ0KyLee** [carla-simulator/carla#7069]
CARLA version: 0.9.15
Platform/OS:Ubuntu 20.04.4LTS
Problem you have experienced:

Following the official documentation steps（https://carla.readthedocs.io/en/latest/build_linux/#set-unreal-engine-environment-variable）, I got to the "make PythonAPI" step and encountered a problem

`...updated 16515 targets...
Setup.sh: Retrieving rpclib.
正克隆到 'rpclib-v2.2.1_c5-c10-source'...
remote: Enumerating objects: 7381, done.
remote: Counting objects: 100% (119/119), done.
remote: Compressing objects: 100% (66/66), done.
remote: Total 7381 (delta 51), reused 82 (delta 39), pack-reused 7262
接收对象中: 100% (7381/7381), 3.37 MiB | 4.39 MiB/s, 完成.
处理 delta 中: 100% (4677/4677), 完成.
注意：正在切换到 '8430cddd7b86a19e19823fac3d72d899794f7256'。

您正处于分离头指针状态。您可以查看、做试验性的修改及提交，并且您可以在切换
回一个分支时，丢弃在此状态下所做的提交而不对分支造成影响。

如果您想要通过创建分支来保留在此状态下所做的提交，您可以通过在 switch 命令
中添加参数 -c 来实现（现在或稍后）。例如：

  git switch -c <新分支名>

或者撤销此操作：

  git switch -

通过将配置变量 advice.detachedHead 设置为 false 来关闭此建议

Setup.sh: Building rpclib with libc++.
CMake Error at /usr/share/cmake-3.16/Modules/CMakeDetermineCCompiler.cmake:49 (message):
  Could not find compiler set in environment variable CC:

  
  /Engine/Extras/ThirdPartyNotUE/SDKs/HostLinux/Linux_x64/v17_clang-10.0.1-centos7/x86_64-unknown-linux-gnu/bin/clang.
Call Stack (most recent call first):
  CMakeLists.txt:2 (project)

CMake Error: CMAKE_C_COMPILER not set, after EnableLanguage
CMake Error: CMAKE_CXX_COMPILER not set, after EnableLanguage
-- Configuring incomplete, errors occurred!
See also "/home/aion/Carla_Lzn/carla/Build/rpclib-v2.2.1_c5-c10-libcxx-build/CMakeFiles/CMakeOutput.log".
make: *** [Util/BuildTools/Linux.mk:142：setup] 错误 1
`

What you expected to happen:  "make PythonAPI“success
Steps to reproduce:
Other information (documentation you consulted, workarounds you tried): I used “sudo make PythonAPI“ instead of "make PythonAPI"

**Comment by nusstu0019** [comment]
I solve this problem by reinstall.

**Comment by viktor-from** [comment]
> I solve this problem by reinstall.

What did you reinstall? Clang or the Carla repo?

**Comment by nusstu0019** [comment]
carla repo
________________________________
发件人: viktorfrom ***@***.***>
发送时间: 2024年2月7日 15:36
收件人: carla-simulator/carla ***@***.***>
抄送: Fan Xiuqi ***@***.***>; Comment ***@***.***>
主题: Re: [carla-simulator/carla] make PythonAPI error : Could not find compiler set in environment variable CC (Issue #7069)

        - External Email -



I solve this problem by reinstall.

What did you reinstall? Clang or the Carla repo?

―

**Comment by yuxinhi** [comment]
I'm facing the same difficulty. Have you resolved this problem ?

**Comment by arashfree** [comment]
I was forgotten to set the correct $UE4_ROOT, set it correct so it would be solved. it should be your installed Unreal folder.

**Comment by LuisPoveda** [comment]
Is this issue been resolved?
Are you fine with closing it?

**Comment by LuisPoveda** [comment]
Since there is no response I am going to close this issue, feel free to reopen it.

**Comment by MOHAMMEDHUSSEIN94** [comment]
> > I solve this problem by reinstall.
> 
> What did you reinstall? Clang or the Carla repo?
> 
> Edit: Seems like my export UE4_ROOT=~/UnrealEngine_4.26 variable in .profile just wasn't set correctly.

I did so and used echo $UE4_ROOT to make sure the directory is that of unreal engine and still getting the same error

**Comment by MOHAMMEDHUSSEIN94** [comment]
> I was forgotten to set the correct $UE4_ROOT, set it correct so it would be solved. it should be your installed Unreal folder.

I did so using gedit ~/.bashrc ,wrote[export UE4_ROOT=~/UnrealEngineCarla] at the end and then ran [source ~/.bashrc] and used echo $UE4_ROOT to make sure the directory is that of unreal engine and still getting the same error

**Comment by YuvarajP** [comment]
I am getting the same error

**Comment by dragnil1** [comment]
I was facing this error while using `sudo make launch`. Firstly, I tried using `make launch`. But got **permission denied** error as my `carla` folder did not have sufficient permission. Then, I enabled read and write permission using the accepted answer of this [question](https://askubuntu.com/questions/487527/give-specific-user-permission-to-write-to-a-folder-using-w-notation). After that, `make launch` worked nicely. Let me know if solution is not appropiate. I also followed [this](https://stackoverflow.com/questions/31421327/cmake-cxx-compiler-broken-while-compiling-with-cmake).

**Comment by JiachengZuo** [comment]
Maybe your clang version is too old; I solved this problem by using these command.
`sudo apt update 
 sudo apt install clang`

**Comment by OYan60718** [comment]
> > I was forgotten to set the correct $UE4_ROOT, set it correct so it would be solved. it should be your installed Unreal folder.
> 
> I did so by writing at the end of bashrc [export UE4_ROOT=~/UnrealEngineCarla] and then ran source and used echo $UE4_ROOT to make sure the directory is that of unreal engine and still getting the same error ![image](https://private-user-images.githubusercontent.com/25697041/326606374-b61be7fc-e4aa-4d20-a2ac-3947abc2ea85.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE5Njk0MjgsIm5iZiI6MTcyMTk2OTEyOCwicGF0aCI6Ii8yNTY5NzA0MS8zMjY2MDYzNzQtYjYxYmU3ZmMtZTRhYS00ZDIwLWEyYWMtMzk0N2FiYzJlYTg1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzI2VDA0NDUyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE0ODgwOWY1NzViN2FkOTQyMDU5OTY1YTFjMTA4YjFkNTgxOWYyYzgwNmViNTVhNDc2OWRhMjQwN2FmM2ViMGYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.kpcWVWNKgKCVaGUNfPwZB-fYdLmU8zRtyRojwqu7OSw)

You need to close and reopen the terminal after setting the path

**Comment by OYan60718** [comment]
> Maybe your clang version is too old; I solved this problem by using these command. `sudo apt update ` ` sudo apt install clang`

Do not try this, this will install the latest version of clang, which is not what CARLA requires.

**Comment by latercoder** [comment]
> I was facing this error while using `sudo make launch`. Firstly, I tried using `make launch`. But got **permission denied** error as my `carla` folder did not have sufficient permission. Then, I enabled read and write permission using the accepted answer of this [question](https://askubuntu.com/questions/487527/give-specific-user-permission-to-write-to-a-folder-using-w-notation). After that, `make launch` worked nicely. Let me know if solution is not appropiate. I also followed [this](https://stackoverflow.com/questions/31421327/cmake-cxx-compiler-broken-while-compiling-with-cmake).

Works for my case!
For some reason, I compiled Carla Client with command "sudo make PythonAPI", then I came across the error "Could not find compiler set in environment variable CC...", then I reinstalled carla from step "git clone ...", and compiled with command "make PythonAPI", then fixed. If you know how to change the permission of carla folder, just try as @dragnil1 did.

**Comment by raynbowy23** [comment]
I spent the whole day trying to make it work. Basically, every past answer sounds acceptable but does not specify enough.

If you encounter this type of error, it means it does not correctly link to your installed Unreal Engine for some reason.
```/Engine/Extras/ThirdPartyNotUE/SDKs/HostLinux/Linux_x64/v17_clang-10.0.1-centos7/x86_64-unknown-linux-gnu/bin/clang```

The correct path should put a path to the Unreal Engine at the front that you are supposed to install already, according to the Carla build document.
```$UE4_ROOT/Engine/Extras/ThirdPartyNotUE/SDKs/HostLinux/Linux_x64/v17_clang-10.0.1-centos7/x86_64-unknown-linux-gnu/bin/clang```.

I tried to set env path in my .bashrc and .bash_profile but the program didn't read it for some reason. Instead, I manually modified the build files and added path, which is UE4_ROOT in front of CC, the path for clang.

In my case specifically, I wrote as below to `Setup.sh`, `BuildOSM2ODR.sh`, `BuidPythonAPI.sh` (maybe I forgot some), which all under the path `/opt/carla/Util/BuildTools/`.
```
export UE4_ROOT="/home/YOUR USERNAME/UnrealEngine_4.26"
export CC="$UE4_ROOT/Engine/Extras/ThirdPartyNotUE/SDKs/HostLinux/Linux_x64/v17_clang-10.0.1-centos7/x86_64-unknown-linux-gnu/bin/clang"
```

Find CC path and add UE4_ROOT. Hope it helps!
